### PR TITLE
Feature/83 password recovery system

### PR DESCRIPTION
**Pull request by Thealean28**
## Description
Added a new password recovery system using 2-factor authentication. If you are logged as an user, you can obtain a qr code to obtain your recovery codes. Then, you can recover your password providing your email associated to your user and the verfication code.

## Related Issue(s)
- #83 

## Checklist
- [x] I have read and followed the project's contributing guidelines.
- [x] I have tested the changes locally.
- [x] I have made corresponding changes to the documentation if necessary.
- [x] My code follows the project's code style guidelines.

## Reviewers
- @gonzcm 

## Tester (if applicable)
- @alblopnov 

## Screenshots 
Using this endpoint, you can obtain your qr code. (Note that you need to be authenticated)
image

Then, you can change your password by filling that form.
image

### CODE REVIEW

**Review by gonzcm — CHANGES_REQUESTED** [review]
The implementation works smoothly. Fix the changes I have mentioned to improve the quality of the code. Great job!

**Comment by gonzcm** [comment]
It would be a good idea to prepare a short briefing so that frontend members understand the implementation, the password recovery flow, and have the qr displayed to the user. This will also help to know if we need to make any changes in the backend. What do you think about it?

File: src/modules/shared/auth/controller.py [hunk 1]
```diff
@@ -52,3 +52,23 @@ async def refresh_controller(db: DataBaseDep, refresh_token: str) -> TokenSchema
...
         "refresh_token": create_refresh_token(user.id),
     }
+
+
+async def get_secret_and_qr(db, email) -> UserSecretOut:
+    user = await find_user_by_email(db, email)
+    if not user:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Correo electrónico no encontrado"
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/auth/controller.py]: Done

File: src/modules/shared/user/router.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
-from fastapi import APIRouter, status
+from fastapi import APIRouter, Depends, status
+from pydantic import EmailStr
 
-from src.core.deps import DataBaseDep
-from src.modules.shared.user.model import UserCreate, UserOut
+from src.core.deps import DataBaseDep, get_current_user
+from src.modules.shared.user import controller
+from src.modules.shared.user.model import User, UserCreate, UserOut
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/user/router.py]: Done

File: src/modules/shared/user/router.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
-from fastapi import APIRouter, status
+from fastapi import APIRouter, Depends, status
+from pydantic import EmailStr
 
-from src.core.deps import DataBaseDep
-from src.modules.shared.user.model import UserCreate, UserOut
+from src.core.deps import DataBaseDep, get_current_user
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/user/router.py]: Done

File: src/modules/shared/user/router.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
-from fastapi import APIRouter, status
+from fastapi import APIRouter, Depends, status
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/user/router.py]: Done

File: src/modules/shared/auth/controller.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/auth/controller.py]: Done

File: src/modules/shared/user/controller.py [hunk 5]
```diff
@@ -13,3 +14,20 @@ async def create_user_controller(db: DataBaseDep, user: UserCreate) -> UserOut:
...
     return result
+
+
+async def change_password_controller(
+        otp_code: str, new_password: str, db: DataBaseDep, email: str) -> dict:
+    user = await service.find_user_by_email(db, email)
+    if not user:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="El correo electrónico introducido no está asociado a ningún usuario")
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/user/controller.py]: Done

File: src/modules/shared/user/controller.py [hunk 6]
```diff
@@ -13,3 +14,20 @@ async def create_user_controller(db: DataBaseDep, user: UserCreate) -> UserOut:
...
+    secret = await get_secret_by_email(db, email)
+    result = verify_otp(secret, otp_code)
+    if not result:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="El código de verificación es incorrecto"
```
[the remark this answers is not in the document]
Comment (Thealean28) [reply src/modules/shared/user/controller.py]: Done

**Review by gonzcm — APPROVED** [review]
Well done!

**Pull request merged**